mibidata/tiff.py

```diff
@@ -82,12 +89,26 @@ def write(filename, image, sed=None, optical=None, ranges=None,
         raise ValueError('Image coordinates and size must not be None.')
     if image.masses is None or image.targets is None:
         raise ValueError('Image channels must contain both masses and targets.')
-    if np.issubdtype(image.data.dtype, np.integer) and not write_float:
-        range_dtype = 'I'
+    if write_float:
+        raise ValueError('`write_float` has been deprecated. Please use the '
+                         '`dtype` argument instead.')
+    if dtype and not dtype in ['float', 'uint16', np.float, np.uint16]:
+        raise ValueError('Invalid dtype specification.')
+    if not dtype:
+        range_dtype = 'I' if\
+            np.issubdtype(image.data.dtype, np.integer) else 'd'
     else:
-        range_dtype = 'd'
+        range_dtype = 'I' if dtype in ['uint16', np.uint16] else 'd'
+
+    save_dtype = np.uint16 if range_dtype == 'I' else np.float
```

Comment:
There was a compatibility reason we were using `np.float32` rather than `float` which is float64. Please retain the use of np.float32 here. Moreover, I think
- we should raise an exception not just a waning if the data isn't equal after converting
- this entire section can be simplified to something like 

```
    if dtype == 'float':
        save_dtype = np.float32
        range_dtype = 'd'
    elif dtype == 'uint16':
        save_dtype = np.uint16
        range_dtype = 'I'
    elif np.issubdtype(image.data.dtype, np.floating):
        save_dtype = np.float32
        range_dtype = 'd'
    else:
        save_dtype = np.uint16
        range_dtype = 'I'

    to_save = image.data.astype(save_dtype)
    if not np.all(np.equal(to_save, image.data)):
        raise ValueError('Cannot convert data from '
                         f'{image.data.dtype} to {save_dtype}')

    if ranges is None:
        ranges = [(0, m) for m in to_save.max(axis=(0, 1))]

Reply:
I like that this forces the data into one of the two types and leaves it on the user to decide how to convert the data before writing the TIFF so that there is no room for confusion later. I elected to keep the check in that `dtype` is one of the allowed to avoid the case where a user specifies `np.uint8` or `abcdef` and not realize why the data is trying to (and failing to be) converted into uint16.

Reply:
100% agreed that there should be a check first that they specify an allowed dtype, my comment didn't make this clear but was intended to be after that check.